src/plugins/dataflow/components/ui/dataflow-program-topbar.tsx

```diff
@@ -86,31 +65,237 @@ export const DataflowProgramTopbar = (props: TopbarProps) => {
+              <PlaybackButton programRecordState={programRecordState}/>
+            }
+          </div>
+          <RateSelectorOrPlayBack
+            rateOptions={programDataRates}
+            dataRate={dataRate}
+            onRateSelectClick={onRateSelectClick}
+            readOnly={readOnly}
+            programRecordState={programRecordState}
+            numNodes={numNodes}
+            onRecordDataChange={onRecordDataChange}
+            finished={finished}
+            handleFinished={handleFinished}
+          />
+          <RecordStopOrClearButton
+            programRecordState={programRecordState}
+            buttonClickHandler={onRecordDataChange}
+          />
+        </div>
+
+        <div className="topbar-right">
+          {props.showRateUI && <span className={"rate-ui"}>{`${props.lastIntervalDuration}ms`}</span>}
+        </div>
+
+    </div>
+  );
+};
+
+/* ==[ Sampling Rate Selector ] == */
+interface IRateSelectorProps {
+  rateOptions: ProgramDataRate[];
+  dataRate: number;
+  onRateSelectClick: (rate: number) => void;
+  readOnly: boolean;
+  programRecordState: number;
+  numNodes: number
+  finished: boolean;
+  handleFinished: (isFinished: boolean)=> void;
+  onRecordDataChange: () => void;
+}
+
+function formatTime(seconds: number) {
+  const minutes = Math.floor(seconds / 60);
+  const remainingSeconds = seconds % 60;
+  const formattedMinutes = minutes.toString().padStart(3, '0');
+  const formattedSeconds = remainingSeconds.toString().padStart(2, '0');
+  return `${formattedMinutes}:${formattedSeconds}`;
+}
+
+const RateSelectorOrPlayBack = (props: IRateSelectorProps) => {
+  const { onRateSelectClick, readOnly, dataRate, rateOptions, programRecordState,
+          numNodes, handleFinished, onRecordDataChange} = props;
+
+  /* ==[ Total Recording Time  - Calculate] format as "MMM:SS" */
+  const totalTimeSec = Math.floor((dataRate / 1000) * (totalSamples/numNodes));
+  const totalTimeFormatted = formatTime(totalTimeSec);
+
+  /* ==[ Timer Recording Time  - Calculate] format as "MMM:SS" */
+  const timerMin = useRef(0);
+  const timerSec = useRef(0); //seconds that have passed after hitting Record
+  const formattedMin = timerMin.current.toString().padStart(3, "0");
+  const formattedSec = timerSec.current.toString().padStart(2, "0");
+  const formattedTime = `${formattedMin}:${formattedSec}`;
+
+  /* ==[ Timer - Enable ] == */
+  const startTimer = numNodes > 0 && timerSec.current < totalTimeSec && (programRecordState === 1);
+
+  useEffect(()=>{
+    const timer = setInterval(() => {
+      startTimer && timerSec.current++;
+      if (timerSec.current === 60){
+        timerMin.current++;
+        timerSec.current = 0;
+      }
+    }, 1000);
+
+    return () => clearInterval(timer);
+  }, [timerSec, startTimer]);
+
+  /* ==[ Timer - Reset ] == */
+  if (programRecordState === 0) {
+    timerSec.current = 0;
+    timerMin.current = 0;
+  }
+
+  if(timerSec.current >= totalTimeSec){
```

Comment:
If `totalTimeSec` is greater than 60, this currently won't work because you reset `timerSec` to 0 every minute.